android-components/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt

```diff
@@ -427,8 +431,15 @@ class SitePermissionsFeature(
             return null
         }
 
-        val private: Boolean = store.state.findTabOrCustomTabOrSelectedTab(sessionId)?.content?.private
-            ?: throw IllegalStateException("Unable to find session for $sessionId or selected session")
+        val private: Boolean? =
+            store.state.findTabOrCustomTabOrSelectedTab(sessionId)?.content?.private
+
+        if (private == null) {
+            logger.error("Unable to find a tab for $sessionId rejecting the prompt request")
+            permissionRequest.reject()
```

Comment:
If the session doesn't exist, what happens when we try to reject this permission from a non-existent session?

Reply:
Ideally we should find a way to improve not crash while we find the root cause(s) :)

Reply:
I followed the code path from the `PermissionRequest.reject()` and found that it goes back to GeckoView to [handle the inner callback][0] and then in the GeckoResult will result in [another exception thrown][1].

So rejecting doesn't seem like the right path.

> That's a good point, and really difficult case to simulate, as we are not completely sure what it's the state of the session on the Gecko side.

(We can pair on this) if we start a private session, open a page that requires an Android permission, then using adb, we kill the content process (there should only be the one tab is that content process), and finally we can try to accept the permission to see what happens.

[0]: https://searchfox.org/mozilla-central/rev/1cae9ca7cc7d17cfc92088dfc68e5ff391128caa/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java#720-731
[1]: https://searchfox.org/mozilla-central/rev/1cae9ca7cc7d17cfc92088dfc68e5ff391128caa/mobile/android/geckoview/src/main/java/org/mozilla/gecko/util/EventCallback.java#48-57

Reply:
Checking the GV side looks like it should be fine [to reject the request](https://searchfox.org/mozilla-central/source/mobile/android/actors/GeckoViewPermissionChild.sys.mjs#65-89) as the whole communication snippet **js->Java** is wrapped by `try` `catch` block. Additionally the session information has been [captured before the permission request is sent to Java](https://searchfox.org/mozilla-central/source/mobile/android/actors/GeckoViewPermissionChild.sys.mjs#58-61).

Something to noticed is that we are doing an [await](https://searchfox.org/mozilla-central/source/mobile/android/actors/GeckoViewPermissionChild.sys.mjs#65) on the response of the permission. 

I'll give a try to and report back :)

Reply:
I found a simpler way that could be closer to the environment where crashes are happening:

1. Apply the below diff on top of this PR.
2. Add break points on `SitePermissionsFeature.kt` from line 437 to 449.
3. Add break points on `TabsRemovedMiddleware` from line 62 to 72 to make sure we are killing the gecko sessions.
4. Run Fenix in debug mode and navigate to https://permission.site/
5. Click the "Notifications" button

On 5 the break point on `SitePermissionsFeature.kt` line 437 will hit.
Let it run as we want to prune all sessions (ac and gecko), then you will start to see break points on `TabsRemovedMiddleware` make sure we un-linked all sessions and let the flow goes.
You will hit `SitePermissionsFeature.kt` on line 442 you will see that `private` is empty then you can proceed. 

So far testing locally I couldn't find any issues please let me know what do you think about this test case.





```diff
diff --git a/android-components/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt b/android-components/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
index 6729a0023a..2d44ac7cfa 100644
--- a/android-components/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
+++ b/android-components/components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt
@@ -15,10 +15,12 @@ import androidx.annotation.VisibleForTesting
 import androidx.fragment.app.FragmentManager
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
 import kotlinx.coroutines.cancel
 import kotlinx.coroutines.flow.distinctUntilChangedBy
 import kotlinx.coroutines.flow.mapNotNull
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
 import kotlinx.coroutines.withContext
 import mozilla.components.browser.state.action.ContentAction
 import mozilla.components.browser.state.action.ContentAction.UpdatePermissionHighlightsStateAction
@@ -32,6 +34,7 @@ import mozilla.components.browser.state.action.ContentAction.UpdatePermissionHig
 import mozilla.components.browser.state.action.ContentAction.UpdatePermissionHighlightsStateAction.MicrophoneChangedAction
 import mozilla.components.browser.state.action.ContentAction.UpdatePermissionHighlightsStateAction.NotificationChangedAction
 import mozilla.components.browser.state.action.ContentAction.UpdatePermissionHighlightsStateAction.PersistentStorageChangedAction
+import mozilla.components.browser.state.action.TabListAction
 import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
 import mozilla.components.browser.state.state.ContentState
 import mozilla.components.browser.state.state.SessionState
@@ -431,6 +434,8 @@ class SitePermissionsFeature(
             return null
         }
 
+        store.dispatch(TabListAction.RemoveAllTabsAction(recoverable = false)).joinBlocking()
+
         val private: Boolean? =
             store.state.findTabOrCustomTabOrSelectedTab(sessionId)?.content?.private
 
@@ -454,6 +459,10 @@ class SitePermissionsFeature(
         return prompt
     }
 
+    fun Job.joinBlocking() {
+        runBlocking { join() }
+    }
+
     @VisibleForTesting
     internal fun handleNoRuledFlow(
         permissionFromStorage: SitePermissions?,
```

Reply:
Nice test case! If you get a chance, consider adding it as a unit test for a general SitePermissionFeature, but I'm not sure myself how to write that.

I checked the logs I do see GeckoView error-ing out gracefully:

```
17398 Gecko                   org.mozilla.fenix.debug              I  console.error: "Permission error:" (new AbortError("Actor 'GeckoViewPermission' destroyed before query 'DispatcherQuery' was resolved", (void 0)))
17398 GeckoConsole            org.mozilla.fenix.debug              E  [JavaScript Error: "Permission error: AbortError: Actor 'GeckoViewPermission' destroyed before query 'DispatcherQuery' was resolved"]
17398 GeckoConsole            org.mozilla.fenix.debug              E  [JavaScript Error: "NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIContentPermissionRequest.cancel]" {file: "resource://gre/modules/GeckoViewPermission.sys.mjs" line: 20}]
```